File: src/platform/ios/GamifiedTaxDeduction/DataModels.swift
import Foundation

// MARK: - User Profile
struct UserProfile: Codable {
    enum IncomeBracket: String, Codable, CaseIterable {
        case low = "low"
        case medium = "medium"
        case high = "high"
    }
    
    enum FilingStatus: String, Codable, CaseIterable {
        case single = "single"
        case married = "married"
    }
    
    let incomeBracket: IncomeBracket
    let filingStatus: FilingStatus
    let createdAt: Date
    let lastUpdated: Date
    
    init(incomeBracket: IncomeBracket, filingStatus: FilingStatus) {
        self.incomeBracket = incomeBracket
        self.filingStatus = filingStatus
        let now = Date()
        self.createdAt = now
        self.lastUpdated = now
    }
}

// MARK: - Receipt
struct Receipt: Codable, Identifiable {
    let id: UUID
    let createdAt: Date
    let lastUpdated: Date
    let imagePath: String
    let vendorName: String?
    let totalAmount: Int // Amount in cents
    let potentialTaxSaving: Int // Tax saving in cents
    let category: String?
    let isVerified: Bool
    let notes: String?
    
    init(imagePath: String, 
         totalAmount: Int,
         potentialTaxSaving: Int,
         vendorName: String? = nil,
         category: String? = nil,
         isVerified: Bool = false,
         notes: String? = nil) {
        self.id = UUID()
        let now = Date()
        self.createdAt = now
        self.lastUpdated = now
        self.imagePath = imagePath
        self.totalAmount = totalAmount
        self.potentialTaxSaving = potentialTaxSaving
        self.vendorName = vendorName
        self.category = category
        self.isVerified = isVerified
        self.notes = notes
    }
    
    // Computed properties for display
    var displayAmount: Double {
        return Double(totalAmount) / 100.0
    }
    
    var displayTaxSaving: Double {
        return Double(potentialTaxSaving) / 100.0
    }
}

// MARK: - Achievement
struct Achievement: Codable, Identifiable {
    enum Category: String, Codable, CaseIterable {
        case scanning = "scanning"
        case savings = "savings"
        case consistency = "consistency"
        case learning = "learning"
    }
    
    let id: UUID
    let name: String
    let description: String
    let category: Category
    let threshold: Int
    let icon: String
    let unlockedAt: Date?
    let isUnlocked: Bool
    
    init(name: String,
         description: String,
         category: Category,
         threshold: Int,
         icon: String) {
        self.id = UUID()
        self.name = name
        self.description = description
        self.category = category
        self.threshold = threshold
        self.icon = icon
        self.unlockedAt = nil
        self.isUnlocked = false
    }
}

// MARK: - User Progress
struct UserProgress: Codable {
    let userId: String
    let totalReceiptsScanned: Int
    let totalPotentialSavings: Int // In cents
    let currentStreak: Int
    let longestStreak: Int
    let lastScanDate: Date?
    let achievementIds: [UUID]
    let createdAt: Date
    let lastUpdated: Date
    
    init(userId: String = "default") {
        self.userId = userId
        self.totalReceiptsScanned = 0
        self.totalPotentialSavings = 0
        self.currentStreak = 0
        self.longestStreak = 0
        self.lastScanDate = nil
        self.achievementIds = []
        let now = Date()
        self.createdAt = now
        self.lastUpdated = now
    }
    
    var displayTotalSavings: Double {
        return Double(totalPotentialSavings) / 100.0
    }
}

// MARK: - Educational Tip
struct EducationalTip: Codable, Identifiable {
    enum Category: String, Codable, CaseIterable {
        case businessMeal = "business_meal"
        case homeOffice = "home_office"
        case equipment = "equipment"
        case general = "general"
    }
    
    struct TriggerConditions: Codable {
        let expenseAmountMin: Int?
        let expenseAmountMax: Int?
        let vendorTypes: [String]?
        let categories: [String]?
        
        init(expenseAmountMin: Int? = nil,
             expenseAmountMax: Int? = nil,
             vendorTypes: [String]? = nil,
             categories: [String]? = nil) {
            self.expenseAmountMin = expenseAmountMin
            self.expenseAmountMax = expenseAmountMax
            self.vendorTypes = vendorTypes
            self.categories = categories
        }
    }
    
    let id: UUID
    let title: String
    let content: String
    let category: Category
    let triggerConditions: TriggerConditions
    let displayCount: Int
    let maxDisplays: Int
    let priority: Int
    
    init(title: String,
         content: String,
         category: Category,
         triggerConditions: TriggerConditions = TriggerConditions(),
         maxDisplays: Int = 3,
         priority: Int = 1) {
        self.id = UUID()
        self.title = title
        self.content = content
        self.category = category
        self.triggerConditions = triggerConditions
        self.displayCount = 0
        self.maxDisplays = maxDisplays
        self.priority = priority
    }
}

// MARK: - Tax Configuration
struct TaxConfiguration: Codable {
    struct IncomeBracket: Codable {
        let min: Int
        let max: Int
        let rate: Double
    }
    
    struct FilingStatusConfig: Codable {
        let standardDeduction: Int
    }
    
    let incomeBrackets: [String: IncomeBracket]
    let filingStatuses: [String: FilingStatusConfig]
    let lastUpdated: Date
    
    static let `default` = TaxConfiguration(
        incomeBrackets: [
            "low": IncomeBracket(min: 0, max: 50000, rate: 0.12),
            "medium": IncomeBracket(min: 50001, max: 100000, rate: 0.22),
            "high": IncomeBracket(min: 100001, max: Int.max, rate: 0.32)
        ],
        filingStatuses: [
            "single": FilingStatusConfig(standardDeduction: 1275000), // $12,750 in cents
            "married": FilingStatusConfig(standardDeduction: 2550000)  // $25,500 in cents
        ],
        lastUpdated: Date()
    )
}

// MARK: - App Settings
struct AppSettings: Codable {
    enum Theme: String, Codable, CaseIterable {
        case light = "light"
        case dark = "dark"
        case system = "system"
    }
    
    let isOnboardingComplete: Bool
    let notificationsEnabled: Bool
    let reminderTime: String?
    let theme: Theme
    let currency: String
    let privacyConsentGiven: Bool
    let lastBackupDate: Date?
    let appVersion: String
    
    static let `default` = AppSettings(
        isOnboardingComplete: false,
        notificationsEnabled: true,
        reminderTime: nil,
        theme: .system,
        currency: "USD",
        privacyConsentGiven: false,
        lastBackupDate: nil,
        appVersion: Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0.0"
    )
}

File: src/platform/android/app/src/main/kotlin/com/gamifiedtaxdeduction/DataModels.kt
package com.gamifiedtaxdeduction.data.model

import androidx.room.*
import java.util.*

// MARK: - User Profile
data class UserProfile(
    val incomeBracket: IncomeBracket,
    val filingStatus: FilingStatus,
    val createdAt: Date = Date(),
    val lastUpdated: Date = Date()
) {
    enum class IncomeBracket {
        LOW, MEDIUM, HIGH
    }
    
    enum class FilingStatus {
        SINGLE, MARRIED
    }
}

// MARK: - Receipt
@Entity(tableName = "receipts")
data class Receipt(
    @PrimaryKey val id: String = UUID.randomUUID().toString(),
    val createdAt: Date = Date(),
    val lastUpdated: Date = Date(),
    val imagePath: String,
    val vendorName: String? = null,
    val totalAmount: Long, // Amount in cents
    val potentialTaxSaving: Long, // Tax saving in cents
    val category: String? = null,
    val isVerified: Boolean = false,
    val notes: String? = null
) {
    // Computed properties for display
    val displayAmount: Double
        get() = totalAmount / 100.0
    
    val displayTaxSaving: Double
        get() = potentialTaxSaving / 100.0
}

// MARK: - Achievement
@Entity(tableName = "achievements")
data class Achievement(
    @PrimaryKey val id: String = UUID.randomUUID().toString(),
    val name: String,
    val description: String,
    val category: Category,
    val threshold: Int,
    val icon: String,
    val unlockedAt: Date? = null,
    val isUnlocked: Boolean = false
) {
    enum class Category {
        SCANNING, SAVINGS, CONSISTENCY, LEARNING
    }
}

// MARK: - User Progress
@Entity(tableName = "user_progress")
data class UserProgress(
    @PrimaryKey val userId: String = "default",
    val totalReceiptsScanned: Int = 0,
    val totalPotentialSavings: Long = 0L, // In cents
    val currentStreak: Int = 0,
    val longestStreak: Int = 0,
    val lastScanDate: Date? = null,
    val achievementIds: List<String> = emptyList(),
    val createdAt: Date = Date(),
    val lastUpdated: Date = Date()
) {
    val displayTotalSavings: Double
        get() = totalPotentialSavings / 100.0
}

// MARK: - Educational Tip
@Entity(tableName = "educational_tips")
data class EducationalTip(
    @PrimaryKey val id: String = UUID.randomUUID().toString(),
    val title: String,
    val content: String,
    val category: Category,
    @Embedded val triggerConditions: TriggerConditions = TriggerConditions(),
    val displayCount: Int = 0,
    val maxDisplays: Int = 3,
    val priority: Int = 1
) {
    enum class Category {
        BUSINESS_MEAL, HOME_OFFICE, EQUIPMENT, GENERAL
    }
    
    data class TriggerConditions(
        val expenseAmountMin: Long? = null,
        val expenseAmountMax: Long? = null,
        val vendorTypes: List<String>? = null,
        val categories: List<String>? = null
    )
}

// MARK: - Tax Configuration
data class TaxConfiguration(
    val incomeBrackets: Map<String, IncomeBracket>,
    val filingStatuses: Map<String, FilingStatusConfig>,
    val lastUpdated: Date = Date()
) {
    data class IncomeBracket(
        val min: Long,
        val max: Long,
        val rate: Double
    )
    
    data class FilingStatusConfig(
        val standardDeduction: Long
    )
    
    companion object {
        val DEFAULT = TaxConfiguration(
            incomeBrackets = mapOf(
                "low" to IncomeBracket(0L, 50000L, 0.12),
                "medium" to IncomeBracket(50001L, 100000L, 0.22),
                "high" to IncomeBracket(100001L, Long.MAX_VALUE, 0.32)
            ),
            filingStatuses = mapOf(
                "single" to FilingStatusConfig(1275000L), // $12,750 in cents
                "married" to FilingStatusConfig(2550000L)  // $25,500 in cents
            )
        )
    }
}

// MARK: - App Settings
data class AppSettings(
    val isOnboardingComplete: Boolean = false,
    val notificationsEnabled: Boolean = true,
    val reminderTime: String? = null,
    val theme: Theme = Theme.SYSTEM,
    val currency: String = "USD",
    val privacyConsentGiven: Boolean = false,
    val lastBackupDate: Date? = null,
    val appVersion: String = "1.0.0"
) {
    enum class Theme {
        LIGHT, DARK, SYSTEM
    }
}

// MARK: - Type Converters for Room Database
class DateConverter {
    @TypeConverter
    fun fromTimestamp(value: Long?): Date? {
        return value?.let { Date(it) }
    }

    @TypeConverter
    fun dateToTimestamp(date: Date?): Long? {
        return date?.time
    }
}

class StringListConverter {
    @TypeConverter
    fun fromStringList(value: List<String>): String {
        return value.joinToString(",")
    }

    @TypeConverter
    fun toStringList(value: String): List<String> {
        return if (value.isEmpty()) emptyList() else value.split(",")
    }
}